### DISCUSSION TITLE
Support type hinting for variables

### DISCUSSION THREAD

**Discussion by saidctb** [pyccel/pyccel#1462] [Ideas]
Adding type hints of variables will allow the user to have more control over the type inference process, allowing the user to add more information to the variable type.
For example:
```python
x: int = 0 # Declare an int and assign zero to it
y: "int32" = 0  # Declare int32 and assign zero to it
z: "int[3,4]" = 0 # Declare a fixed size array
w: "Final" = 0 # Declare a constant int and assign zero to it
t: "Final[int32]" = 0 # Declare a constant int32 and assign zero to it

```

**Comment by EmilyBourne** [comment]
Adding support for this would require the addition of a `_visit_AnnAssign` node in the `SyntacticParser`.
I see two possible ways to manage the addition of this:
1. Add a new type to handle this. E.g an `AnnotatedPyccelSymbol` which inherits from `PyccelSymbol`
2. Create the `Variable` directly in the syntactic stage. This blurs the lines between the syntactic and semantic stage which makes it somewhat dangerous but it avoids unnecessary operations

**Reply by saidctb** [reply]
The first option is better in my opinion

**Comment by EmilyBourne** [comment]
Apart from the addition of a "Final" ("const") annotation, can you explain what this would add that the user couldn't achieve in pure Python? E.g:
```
@stack_array('z')
def f():
    x = 0
    y = np.int32(0)
    z = np.zeros((3,4))
```
I am generally not a fan of handling things in annotations which can be done in pure Python as it leads to the generated code behaving differently to the original Python code. As a result even if we were to support syntax such as:
```
y: "int32" = 0
```
I would expect it to raise a warning as a minimum to show that the results may not be as expected after testing in Python

**Reply by saidctb** [reply]
I agree that warnings should be raised if the results are not the same as Python.
recently I wrote a for loop and I wanted the index to be of type int32 and
the only way to do it is by doing this
```python
from numpy import int32
i = int32(0)
for i in range(10):
    ...
```

so I thought of having type annotations as a way to solve this.
another use case is in the stub files which is related to the issue [641](https://github.com/pyccel/pyccel/issues/641) .
another use case is to specify that a variable is positive or negative, this will allow us to do some optimizations.

**Reply by EmilyBourne** [reply]
Your example can also be achieved with:
```
import numpy as np
for i in np.arange(10, dtype = np.int32):
     ....
```
Although I agree that this is sub-optimal as `arange` allocates memory and `range` doesn't.

It is interesting that your solution works, it probably ought to raise an error if we are completely strict with the typing as in Python the type of `i` would change.

An alternative which better matches Python would be:
```
import numpy as np
for idx in range(10):
    i = np.int32(idx)
     ....
```

**Reply by EmilyBourne** [reply]
The use case in stub files is a very good point

**Reply by saidctb** [reply]
I wanted to create many versions of the same function with different precisions, a quick way to do it is by using templates, it won't work in the way you have suggested. We also occasionally have many for-loops and the code will be messy if each time we change the precision of the index in the manner you have suggested.

**Reply by EmilyBourne** [reply]
I'm not sure how templates make a difference? Can you explain why it won't work?

I understand the argument for multiple loops though. although there is no difference/gain between:
```
i = int32(0)
```
and
```
i : int32 = 0
```

**Reply by saidctb** [reply]
you can see the advantage when we introduce templates, this will give us the possibility to have many possible combinations 
for example, we can't have a function that does this
```python
def f(a:int32):
    x = np.int32(0) or np.int64(0)
```
using type annotations we can do
```python
@template(name='T', types=['int32','int64'])
def f(a:T):
    x : T = 0
```

**Reply by EmilyBourne** [reply]
This is a reoccurring issue. I agree the type hinting needs tidying up to allow us to support this or:
```
@template(name='T', types=['int32','int64'])
def f(a:T):
    x = T(0)
```

**Reply by saidctb** [reply]
I'm not sure if we should treat types as functions, especially that we can have array types, and for sure the code won't run on Python

**Reply by saidctb** [reply]
another use case is when we deal with the array orders using templates we can have something like
```python
@template(name='T', types=['int32[:,:](order=C)','int64[:,:](order=F)'])
def f(a:T):
    x : T = np.zeros((2,3))
```
maybe there is a cleaner way of doing it but at the moment this seems to me the easiest way of doing it

**Reply by EmilyBourne** [reply]
I was thinking of them as classes. The code could run on Python with the correct definition of `template` but I am not sure that it would be as simple as I would hope.

**Reply by EmilyBourne** [reply]
> maybe there is a cleaner way of doing it but at the moment this seems to me the easiest way of doing it

I think it would be cleaner to write this as:
```
x = np.zeros_like(a, shape=(2,3))
```
